### DISCUSSION TITLE
Does Hamilton support datasets that are split across multiple files, specifically non-tabular data?

### DISCUSSION THREAD

**Discussion by mtillman14** [apache/hamilton#1195] [Q&A]
I'm a Ph.D. student in biomechanics doing lots of data analysis. My datasets consist of data from multiple Subjects that each have multiple Trials. Each Trial's binary data file lives in the `Subject/Trial` folder. Unlike tabular data where a column (a variable) is often processed all at once (from one or multiple CSV files), which seems to be how Hamilton functions (couldn't find this in the docs) in my dataset that is not possible. In my dataset, each "row" is one Trial, which contains multiple timeseries variables (for example, gait speed over time, or even multiple timeseries such as multiple force sensors' readings over time). Therefore each Trial must be processed separately.

Is this structure of dataset & analysis supported by Hamilton? This is such a cool tool and I really hope to be able to use it.

**Comment by skrawcz** [comment]
Thanks for the questions @mtillman14 .

In short Hamilton is data type agnostic. So I'm sure there's a way to get what you want. But to help you some questions:

1. Can you share a sketch of your data so that we're on the same page please? An example of the structure would be great.
2. What would you like to do with the data? What types of transformations? Do you have any current code now that you could share to give us a sense of that?

**Reply by mtillman14** [reply]
Thanks for entertaining the discussion!
1. I work with motion capture data, which generates .c3d binary files, one per trial. In each .c3d file is essentially a dictionary of 3D position data, containing multiple markers' position data. The dictionary keys are marker names, the values are the 3D coordinates of marker positions (N x 3, where N is the length of the timeseries).

The .c3d files are arranged in my file system like this:
./
|--- Subject1/
|---|--- Trial1.c3d
|---|--- Trial2.c3d
|--- Subject2/
|---|--- Trial1.c3d
|---|--- Trial2.c3d

2. My data processing pipeline has like 80 individual steps. Each step is performed on each Trial of each Subject, like this pseudocode example where I compute the length of a person's steps in each Trial:
```
for subject in subjects:
    for trial in subjects[subject]:
        # Here I load data, process it, and save data. 
        # Trivial pseudocode example:
        my_trial_data = load_data([subject, trial])
        trial_step_length_data = compute_step_length(my_trial_data)
        save_data(my_data_squared, [subject, trial])
```

The high level steps in my pipeline are:
- Calibrate a person-specific biomechanical model from marker data
      - Note that this step requires referencing calibration trials (for example when analyzing e.g. Trial 3, maybe I need to reference a calibration performed in Trial 1).
- Use that model to compute metrics like center of mass position/velocity/acceleration, step length, etc.
      - These steps are much more straightforward, it's just math
- Plot the data and run descriptive statistics & hypothesis testing

**Reply by skrawcz** [reply]
Cool. To clarify:
> In each .c3d file is essentially a dictionary of 3D position data, containing multiple markers' position data. The dictionary keys are marker names, the values are the 3D coordinates of marker positions (N x 3, where N is the length of the timeseries).

So this would be:
```python
{"marker1": [(1, 1, 1), (1, 1, 2), ....]} 
```
Is there any time information? or?

In terms of processing this, it sounds like there would be two stages:

1. transform data into tabular representation (I assume this is more useful for then doing analysis etc?):
For a trial create something like:
time column |   marker1 | marker 2 | marker 3| ...
time 0.       | (1,1, 1)  | (3,2,2) |  ...
time 1        |  (1, 1, 2) | ...
time 2       | ...

2. then perform transforms.

Is that roughly how you're thinking about it?

**Reply by mtillman14** [reply]
The dict you wrote is correct, that's the format of our marker data. Timestamps are not included by default, though some people choose to add them in as you have done.

We generally use the marker data dict directly, there's no benefit for us in converting to tabular form. For example, the first thing we generally do is smooth the marker data. For that, we just iterate over each marker name's key, apply the smoothing, and save the smoothed data in another dict. This is then repeated for each Trial of each Subject.

Does this lend itself to a Hamilton-based workflow?

**Reply by skrawcz** [reply]
> We generally use the marker data dict directly, there's no benefit for us in converting to tabular form.

Really? Oh okay. I assume this is "small" data then? Tabular data just means you can make better use of underlying data structures/types, e.g. pyarrow, to do faster computation - e.g. you can then also leverage their implementations, e.g. smoothing via pandas API, etc.

> For example, the first thing we generally do is smooth the marker data. For that, we just iterate over each marker name's key, apply the smoothing, and save the smoothed data in another dict. 

Okay so you basically for each operation create a new dictionary? Sure that works.

Then a sketch of code would be something like this -- assuming you do everything to every marker and treat that data set as a "whole" if you get what I mean:

```python
def raw_marker_data(subject: str, trial: int) -> dict:
    # load marker data for subject trial
    return # some dict

def smoothed_data(raw_marker_data: dict, some_value: int) -> dict:
   # code to smooth ; some_value is some value for smoothing
   return # new dict

def next_transform(smoothed_data: dict) -> dict:
  # code for next transform
  return # new dict
```
would then create something like this:
![Screen Shot 2024-10-19 at 10 14 46 AM](https://github.com/user-attachments/assets/5f739015-7a71-423f-b46f-5dff397e8226)

Then when you come to run things, you'd just request what "outputs" you want and Hamilton would only walk that part of the graph.

**Reply by mtillman14** [reply]
This all looks sensible for pipeline dependency management within one Subject & Trial, I'm just not sure how it's implemented when running the code across multiple Subjects & Trials? If I have 10 Subjects each with 10 Trials (100 Trials total), how would Hamilton help manage that?

One of my biggest needs for pipeline management is for example when execution is interrupted during a particular Trial (e.g. Subject 2 Trial 5), and after fixing whatever issue caused the interruption I want to resume processing from that same Trial (Subject 2 Trial 5) onward. Ideally, my management tool would automagically know to resume from Subject 2 Trial 5, instead of restarting at Subject 1 Trial 1. 

Does Hamilton support this? Thank you!

**Reply by skrawcz** [reply]
> This all looks sensible for pipeline dependency management within one Subject & Trial, I'm just not sure how it's implemented when running the code across multiple Subjects & Trials? If I have 10 Subjects each with 10 Trials (100 Trials total), how would Hamilton help manage that?

You could either do the double for loop you had, and delegate the inside to Hamilton. Or you use a construct like [Parallelizable/Collect](https://hamilton.dagworks.io/en/latest/concepts/parallel-task/).

> One of my biggest needs for pipeline management is for example when execution is interrupted during a particular Trial (e.g. Subject 2 Trial 5), and after fixing whatever issue caused the interruption I want to resume processing from that same Trial (Subject 2 Trial 5) onward. Ideally, my management tool would automagically know to resume from Subject 2 Trial 5, instead of restarting at Subject 1 Trial 1.

So Hamilton doesn't contain a database. You would need to use something to manage what was run, e.g. after each successful run mark it somewhere in the filesystem. You can then check that say in the [Parallelizable function to know whether to retry it or not](https://hamilton.dagworks.io/en/latest/concepts/parallel-task/). This used in conjunction with the [new caching functionality](https://blog.dagworks.io/p/feature-release-caching?r=2cg5z1&utm_campaign=post&utm_medium=web) would be a lightweight way to achieve what you want; the caching would only rerun parts that would have changed.

The Hamilton UI would be useful in tracking / seeing runs and executions. But we haven't exposed querying that for a purpose such as this.

CC @zilto who can provide some more thoughts here next week.

**Reply by mtillman14** [reply]
Thank you, this is great info! My takeaway so far is that the best way to use Hamilton in my workflow would be to treat each unique Subject & Trial combination as its own individual "run" in the Hamilton UI? Is there a way to have something like "nested runs" so that I could see that the last 100 "runs" were all part of 1 meta-run across all 100 Trials? Maybe even have summary stats for the meta-run?

**Reply by skrawcz** [reply]
So with [Parallelizable/Collect](https://hamilton.dagworks.io/en/latest/concepts/parallel-task/) you could group a bunch of trials into a single nested run but it would be seen as one big run. Note if any failures occurred, it would stop the whole run.

Alternatively you can "tag" runs -- so you could then filter to them all in the UI. e.g. tag the subject and/or trial, the for a given subject see their trials. This would give you some visibility here -- but we could definitely add more functionality here.

Office hours are on Tuesdays 9:30am PT if you wanted to drop by -- link in the slack.
